document/AnnexE-Determinism

```diff
@@ -0,0 +1,136 @@
+(because builds performed on different machines might use a different root directory)
+* `annotatedCodeLocation.threadId`
+* `notification.threadId`
+* `notification.time`
+* `result.baselineState` (because it depends on a previous run.
+A result whose `baselineState` was `"new"` in a previous run would be `"existing"`
+in the next run.)
+* `run.id`
+* `run.automationId`
+* `run.baselineId`
+* `stackFrame.threadId`
+* `stackFrame.address`
+* The use of absolute file paths in URI-valued properties such as `physicalLocation.uri`
+* The presence of any non-deterministic elements in a property bag property
+
+## Array and dictionary element ordering { #Annex-E-Ordering }
+
+For a tool to produce deterministic output, it must emit array and dictionary elements
+in a deterministic order.
+
+For some arrays, the SARIF format requires a specific ordering.
+For example, within the `stack.Frames` property, SARIF requires the `annotatedCodeLocation` object
+representing the most deeply nested function call to appear first.
+
+For other arrays, the SARIF format does not require a specific ordering.
+For example, within the `file.hashes` property, SARIF does not require the `hash` objects
+to appear in any particular order.
+For such arrays, a tool can ensure the order by sorting the array elements before
+writing them to the log file.
+For example, it might sort the `hash` objects alphabetically by the string value of the
+`hash.algorithm` property.
+
+A tool might similarly choose to emit the string elements of a `properties.tags` array in
+locale-insensitive alphabetical order.
+
+The array of `result` objects presents more of a problem.
+A multi-threaded analysis tool analyzing multiple files in parallel might produce results
+in any order, and there is no natural order for the results.
+A tool might choose to order them, for example, first alphabetically by analysis target URI,
+then numerically by line number, then by column number, then alphabetically by rule id.
+
+For dictionaries such as the `run.rules` object or the `run.files` object,
+a tool might order the property names alphabetically, using a locale-insensitive ordering.
+
+## Case-insensitive file format elements { #Annex-E-Case }
+
+Case-insensitive file systems can also cause non-deterministic output.
+Suppose a URI-valued property contains a file path from a file system that is
+case-preserving but case-insensitive (such as NTFS):
+
+```
+"uri": "file:///C:/Code/File1.c"
+```
+
+Suppose the file is then renamed to `file.c`. NTFS treats `File.c` and `file.c` as the same file.
+But a subsequent run of the tool might produce
+
+```
+"uri": "file:///C:/Code/file1.c"
+```
+
+A tool might choose to ignore this issue, since this would happen rarely, and the only
+consequence of ignoring it is that a subsequent step in a build process might execute
+even though it does not strictly have to.
+
+To produce absolutely deterministic output, the tool might instead choose to pull all
```

Comment:
Changed to "convert"/

---
In reply to: [67000883](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r67000883) [](ancestors = 67000883)